src/args.rs

```diff
@@ -0,0 +1,441 @@
+		impl FromStr for ParsedRange {
+			type Err = ParseIntError;
+
+			fn from_str(s: &str) -> Result<Self, Self::Err> {
+				let range = match s.split_once('-') {
+					Some((start, end)) => start.parse()?..=end.parse()?,
+					None => {
+						let idx = s.parse()?;
+						idx..=idx
+					}
+				};
+				Ok(Self(range))
+			}
+		}
+
+		ranges
+			.into_iter()
+			.map(ParsedRange::from_str)
+			.flat_map(|range| match range {
+				Ok(range) => Either::Left(range.0.map(Ok)),
+				Err(err) => Either::Right(iter::once(Err(err))),
+			})
+	}
+
+	fn parse_ranges(ranges: &str) -> Result<Vec<usize>, ParseIntError> {
+		Self::parse_ranges_iter(ranges.split([' ', ','].as_slice())).collect()
+	}
+}
+
+#[derive(Error, Debug)]
+pub enum ParseAffinityError {
+	#[error(transparent)]
+	Parse(#[from] ParseIntError),
+
+	#[error("Available cores: {available_cores:?}, requested affinities: {requested_affinities:?}")]
+	InvalidValue {
+		available_cores: Vec<usize>,
+		requested_affinities: Vec<usize>,
+	},
+}
+
+impl FromStr for Affinity {
+	type Err = ParseAffinityError;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		let available_cores = core_affinity::get_core_ids()
+			.unwrap()
+			.into_iter()
+			.map(|core_id| core_id.id)
+			.collect::<Vec<_>>();
+
+		let requested_affinities = Self::parse_ranges(s)?;
+
+		if !requested_affinities
+			.iter()
+			.all(|affinity| available_cores.contains(affinity))
+		{
+			return Err(ParseAffinityError::InvalidValue {
+				available_cores,
+				requested_affinities,
+			});
+		}
+
+		let core_ids = requested_affinities
+			.into_iter()
+			.map(|affinity| CoreId { id: affinity })
+			.collect();
+		Ok(Self(core_ids))
+	}
+}
+
+impl<'de> Deserialize<'de> for Affinity {
+	/// Takes a list of core IDs and attempts to iteratively deserialize them.
+	///
+	/// **Note:** Heavily inspired from https://serde.rs/impl-deserialize.html
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: Deserializer<'de>,
+	{
+		struct AffinityVisitor;
+
+		impl<'de> Visitor<'de> for AffinityVisitor {
+			type Value = Affinity;
+
+			fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+				formatter.write_str(
+					"a list of core IDs (usize) or range of core IDs (String, e.g. '1-3')",
+				)
+			}
+
+			fn visit_seq<A>(self, mut seq: A) -> Result<Affinity, A::Error>
+			where
+				A: SeqAccess<'de>,
+			{
+				/// In a TOML configuration file, strings containing ranges or digits (e.g. '4-5', '4'),
+				/// as well as ordinary digits (e.g. 4) are supported.
+				///
+				/// This means that a combination such as ['4-5', '6', 7] is allowed in the TOML config.
+				/// This is a byproduct of the preexisting clap interface the FromStr trait implementation
+				/// of Affinity, but there wasn't too much of a reason to restrict it; we can support both,
+				/// but merely document combinations like ['4-5', 6] instead.
+				#[derive(Deserialize)]
+				#[serde(untagged)]
+				enum AffinityInputTypes {
```

Comment:
Suggestion: Do something like https://gist.github.com/fogti/9dd33c564f95a9bf6a655d715048c789 instead.

Reply:
I put this into #1036.